cvat/apps/engine/cloud_provider.py

```diff
@@ -196,29 +206,30 @@ def optimally_image_download(self, key: str, chunk_size: int = 65536) -> BytesIO
     def bulk_download_to_memory(
         self,
         files: List[str],
-        threads_number: int = min(get_cpu_number(), 4),
+        *,
+        threads_number: Optional[int] = None,
         _use_optimal_downloading: bool = True,
-    ) -> List[BytesIO]:
+    ) -> Iterator[BytesIO]:
         func = self.optimally_image_download if _use_optimal_downloading else self.download_fileobj
-        if threads_number > 1:
-            with ThreadPool(threads_number) as pool:
-                return pool.map(func, files)
-        else:
-            slogger.glob.warning('Download files to memory in series in one thread.')
-            return [func(f) for f in files]
+        threads_number = normalize_threads_number(threads_number, len(files))
+
+        with ThreadPoolExecutor(max_workers=threads_number) as executor:
+            yield from executor.map(func, files)
 
     def bulk_download_to_dir(
         self,
         files: List[str],
         upload_dir: str,
-        threads_number: int = min(get_cpu_number(), 4),
-    ):
+        *,
+        threads_number: Optional[int] = None,
+    ) -> None:
+        threads_number = normalize_threads_number(threads_number, len(files))
+
         args = zip(files, [os.path.join(upload_dir, f) for f in files])
         if threads_number > 1:
             with ThreadPool(threads_number) as pool:
-                return pool.map(lambda x: self.download_file(*x), args)
+                pool.map(lambda x: self.download_file(*x), args)
```

Comment:
Probably, you need to wait for the results of the call. The function returns futures to be awaited without waiting for their completion. A simple way is to add a `list()` call for the map result.

https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.map